mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h

```diff
@@ -235,6 +235,11 @@ void populateVectorExtractStridedSliceToExtractInsertChainPatterns(
     std::function<bool(ExtractStridedSliceOp)> controlFn = nullptr,
     PatternBenefit benefit = 1);
 
+/// Pattern to rewrite simple cases of N-D extract_strided_slice, where the
+/// slice is contiguous, into extract and shape_cast.
+void populateVectorContiguousExtractStridedSliceToExtractPatterns(
+    RewritePatternSet &patterns, PatternBenefit benefit = 1);
```

Comment:
Can you expand on why isn't this a good canonicalization?

Reply:
Two things (that happened in the testcase I looked at, where these ops where extracting parts from a matrix tile to feed into GPU matrix multiplication intrinsics):
1. `extract` is more constrained than `extract_strided_slice`, so it is more likely to have a good lowering.
2. my use case was, similarly to the test added in this PR, a `extract_strided_slice` producing a vector with leading unit dims, followed by a `shape_cast` dropping the unit dims. That `shape_cast` was hitting the fallback lowering pattern, `ShapeCastOpRewritePattern`. Now that the `extract_strided_slice` is rewritten into a pair (`extract`, `shape_cast`), the two `shape_cast` fold.

Reply:
Sorry @joker-eph, I mis-parsed your question --- read "is"  instead of "isn't".

No opinion about whether this should be a "canonicalization".  I wasn't too sure that I wanted to enter that debate; my pattern is replacing 1 op with 2 ops so I expected a nontrivial debate. Feel free!

Reply:
I would like this kind of consideration of be carefully done **before** adding random patterns to the codebase.

I am really concerned about the lack of design coming with adding single pattern with single "populateXXX" methods. This can't scale and does not help defining a cohesive system.

Reply:
Yeah individual patterns and populate is a problem. The proliferation of such methods is difficult to keep track of.

1. I dont have enough to go by whether this is a canonicalization or not. In absence of evidence, and (at least some) justification, id bias towards not a canonicalization. Basically question is if this pattern will unlock other canonicalizations, then its a strong signal that this is a canonicalization.
2. @joker-eph whats your solution for such cases. This kind of thing is a very point-fix for something. There is really not a over-arching problem being solved really. So one option is, we dont add it to MLIR itself, but keep it downstream. That will only make such things be duplicated across all MLIR projects. Where/how do we house such one-off things in core.

Reply:
I'll step aside and let the compiler experts chime in! I just needed an ad-hoc pattern to unblock my project. I thought that having it upstream rather than downstream was a healthy default to have, and didn't have to be a lot more work than downstream; and if the correct next step is to debate whether this pattern is a canonicalization, then hopefully having the concrete code to look at helps to root this conversation. No strong opinion though and I'll happily go with what others decide here, for next time.

Reply:
>  Basically question is if this pattern will unlock other canonicalizations,

I don't think this is a necessary criteria, for example @bjacob above mentioned:

- extract is more constrained than extract_strided_slice, so it is more likely to have a good lowering.
- a extract_strided_slice producing a vector with leading unit dims, followed by a shape_cast dropping the unit dims. That shape_cast was hitting the fallback lowering pattern, ShapeCastOpRewritePattern. 


These are enough of a motivation to justify a canonicalization to me. Then on top of this is the question of whether the transformation is potentially losing semantics that can't be trivially reconstructed (such aspect would likely make it clearly not suitable for canonicalization).

> @joker-eph whats your solution for such cases. 

Make it a canonicalization.

> So one option is, we dont add it to MLIR itself, but keep it downstream. 

If this can't be grouped in a cohesive pass that achieve something meaningful that we can reason about, then yeah please keep all these patterns out-of-tree.

Reply:
> If this can't be grouped in a cohesive pass that achieve something meaningful that we can reason about, then yeah please keep all these patterns out-of-tree.

At the risk of duplication in all downstream projects? I am fine with that if that is the consensus, but IMO many times, the "pattern" across patterns only appears when they are all put in the same place, i.e. it is not always possible to have overarching plans for everything from the get go, but rather you build that intuition over time.

Reply:
> At the risk of duplication in all downstream projects?

I don't have much concerns about this: if it is really valuable, then it may get upstreamed properly eventually. We don't need to take on any random pattern without organization or rationale, just because it fitted some particular downstream flow.

> but IMO many times, the "pattern" across patterns only appears when they are all put in the same place

This does not really match the way I approach the design. Do you have prior example of success story that would support this?

Reply:
> This does not really match the way I approach the design. Do you have prior examples of success story that would support this? Right now I see an abuse of populateXXXPattern without much convergence.

Not for patterns, but has been the case for a couple of Interfaces that were initially just patterns that did similar things and then were consolidated to use interface.

Reply:
Good discussion! To help keep it concrete, I gave the canonicalizer idea a try: https://github.com/llvm/llvm-project/pull/111614.